README.md
# EVM-LITE

[![CircleCI](https://circleci.com/gh/mosaicnetworks/evm-lite.svg?style=svg)](https://circleci.com/gh/mosaicnetworks/evm-lite)
[![Go Report](https://goreportcard.com/badge/github.com/mosaicnetworks/evm-lite)](https://goreportcard.com/report/github.com/mosaicnetworks/evm-lite)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

## A lean Ethereum node with interchangeable consensus.

We took the [Go-Ethereum](https://github.com/ethereum/go-ethereum)
implementation (Geth) and extracted the EVM and Trie components to create a lean
and modular version with interchangeable consensus.

The EVM is a virtual machine specifically designed to run untrusted code on a
network of computers. Every transaction applied to the EVM modifies the State
which is persisted in a Merkle Patricia tree. This data structure allows to
simply check if a given transaction was actually applied to the VM and can
reduce the entire State to a single hash (merkle root) rather analogous to a
fingerprint.

The EVM is meant to be used in conjunction with a system that broadcasts
transactions across network participants and ensures that everyone executes the
same transactions in the same order. Ethereum uses a Blockchain and a Proof of
Work consensus algorithm. EVM-Lite makes it easy to use any consensus system,
including [Babble](https://github.com/mosaicnetworks/babble) .

## ARCHITECTURE

```
                +-------------------------------------------+
+----------+    |  +-------------+         +-------------+  |       
|          |    |  | Service     |         | State       |  |
|  Client  <-----> |             | <------ |             |  |
|          |    |  | -API        |         | -EVM        |  |
+----------+    |  |             |         | -Trie       |  |
                |  |             |         | -Database   |  |
                |  +-------------+         +-------------+  |
                |         |                       ^         |     
                |         v                       |         |
                |  +-------------------------------------+  |
                |  | Engine                              |  |
                |  |                                     |  |
                |  |       +----------------------+      |  |
                |  |       | Consensus            |      |  |
                |  |       +----------------------+      |  |
                |  |                                     |  |
                |  +-------------------------------------+  |
                |                                           |
                +-------------------------------------------+

```

## Usage

EVM-Lite is a Go library, which is meant to be used in conjunction with a 
consensus system like Babble, Tendermint, Raft, etc.

This repo contains **Solo**, a bare-bones implementation of the consensus 
interface, which is used for testing or launching a standalone node. It relays
transactions directly from Service to State.

## Configuration

The Ethereum genesis file defines Ethereum accounts and is stripped of all the 
Ethereum POW stuff. This file is useful to predefine a set of accounts that own 
all the initial Ether at the inception of the network.

Example Ethereum genesis.json defining two account:
```json
{
   "alloc": {
        "629007eb99ff5c3539ada8a5800847eacfc25727": {
            "balance": "1337000000000000000000"
        },
        "e32e14de8b81d8d3aedacb1868619c74a68feab0": {
            "balance": "1337000000000000000000"
        }
   }
}
```
## API

The Service exposes an HTTP API.  

### Get Account

Retrieve information about any account.  

```bash
host:~$ curl http://[api_addr]/account/0x629007eb99ff5c3539ada8a5800847eacfc25727 -s | json_pp
{
    "address": "0xa10aae5609643848fF1Bceb76172652261dB1d6c",
    "balance": 1234567890000000000000,
    "nonce": 0,
    "bytecode": ""
}
```

### Call

Call a smart-contract READONLY function. These calls will NOT modify the EVM
state, and the data does NOT need to be signed.

```bash
curl http://localhost:8080/call \
    -d '{"constant":true,"to":"0xabbaabbaabbaabbaabbaabbaabbaabbaabbaabba","value":0,"data":"0x8f82b8c4","gas":1000000,"gasPrice":0,"chainId":1}' \
    -H "Content-Type: application/json" \
    -X POST -s | json_pp
    {
      "data": "0x0000000000000000000000000000000000000000000000000000000000000001"
    }
```

### Submit Transaction

Send a SIGNED, NON-READONLY transaction. The client is left to compose a
transaction, sign it and RLP encode it. The resulting bytes, represented as a
Hex string, are passed to this method to be forwarded to the EVM. This is a
SYNCHRONOUS operation; it waits for the transaction to go through consensus and
returns the transaction receipt.

example:
```bash
host:~$ curl -X POST http://[api_addr]/rawtx -d '0xf86904808398968094f7cd2ba6892341e568e9d825c4bdc2bd53b7524189031b9d1340ad2500008026a04eb7420aa52a1955d26ffb16d3a8cb8d969ae0eb6d75bb5076599c42a788e08da0178b3ddb264cdcc624121f55a95ae45de119bc44a0a85b721d8958b7ebe0553a' -s | json_pp
{
   "root": "0xda4529d2bc5e8b438edee4463637eb91d5490edb50d15e786e8d5276f2a2c8f4",
   "transactionHash": "0x3f5682786828d26946e12a08a858b6dd805d1ea8f7d39d93f1d4d5393b23f710",
   "from": "0x888980abf63d4133482e50bf8233f307e3c2b941",
   "to": "0xf7cd2ba6892341e568e9d825c4bdc2bd53b75241",
   "gasUsed": 21000,
   "cumulativeGasUsed": 21000,
   "contractAddress": "0x0000000000000000000000000000000000000000",
   "logs": [],
   "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
   "status": 1
   }
```

### Get Transaction Receipt

Get a transaction receipt. When a transaction is applied to the EVM, a receipt
is saved to record if/how the transaction affected the state. This contains
such information as the address of a newly created contract, how much gas was
use, and the EVM Logs produced by the execution of the transaction.

example:
```bash
host:~$ curl http://[api_addr]/tx/0xeeeed34877502baa305442e3a72df094cfbb0b928a7c53447745ff35d50020bf -s | json_pp
{
   "to" : "0xe32e14de8b81d8d3aedacb1868619c74a68feab0",
   "root" : "0xc8f90911c9280651a0cd84116826d31773e902e48cb9a15b7bb1e7a6abc850c5",
   "gasUsed" : "0x5208",
   "from" : "0x629007eb99ff5c3539ada8a5800847eacfc25727",
   "transactionHash" : "0xeeeed34877502baa305442e3a72df094cfbb0b928a7c53447745ff35d50020bf",
   "logs" : [],
   "cumulativeGasUsed" : "0x5208",
   "contractAddress" : null,
   "logsBloom" : "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"
}

```

## Info

The `/info` endpoint exposes a map of information provided by the consensus
system.

example (with Babble consensus):
```bash
host:-$ curl http://[api_addr]/info | json_pp
{
   "rounds_per_second" : "0.00",
   "type" : "babble",
   "consensus_transactions" : "10",
   "num_peers" : "4",
   "consensus_events" : "10",
   "sync_rate" : "1.00",
   "transaction_pool" : "0",
   "state" : "Babbling",
   "events_per_second" : "0.00",
   "undetermined_events" : "22",
   "id" : "1785923847",
   "last_consensus_round" : "1",
   "last_block_index" : "0",
   "round_events" : "0"
}

```

## CLIENT

Please refer to [EVM-Lite CLI](https://github.com/mosaicnetworks/evm-lite-cli)
for Javascript utilities and a CLI to interact with the API.

## DEV

DEPENDENCIES

We use glide to manage dependencies:

```bash
[...]/evm-lite$ curl https://glide.sh/get | sh
[...]/evm-lite$ glide install
```
This will download all dependencies and put them in the **vendor** folder; it
could take a few minutes.


CHANGELOG.md

# Changelog

## Unreleased

SECURITY:
FEATURES:
IMPROVEMENTS:
BUG FIXES:

## v0.3.7 (November 27, 2019)

BUG FIXES:

- state: set account storage and nonce when creating accounts

## v0.3.6 (November 7, 2019)

IMPROVEMENTS:

- service: new export function and endpoint that returns a full snapshot of the
           state, which can be reused as a genesis file. 

## v0.3.5 (October 15, 2019)

IMPROVEMENTS:

- state: optimisations and performance tuning

## v0.3.4 (September 18, 2019)

IMPROVEMENTS:

- state: more granular use of mutexes.
- service: higher throughput thanks to above improvement.

## v0.3.3 (September 13, 2019)

FEATURES:

- currency: new denominations for token units

BUG FIXES:

- state: handling transaction promises and errors

## v0.3.2 (September 6, 2019)

FEATURES:
- state: make use of coinbase address
- service: min gas price

## v0.3.1 (August 29, 2019)

SECURITY:
- service: enable CORS

IMPROVEMENTS:
- service: make /tx synchronous (directly return receipt)

## v0.3.0 (August 8, 2019)

Refactor. EVM-Lite becomes a library

SECURITY:
- service: removed keystore
- service: removed /tx endpoint for unsigned txs
- service: removed CORS

FEATURES:
- service: handlers registered to http.DefaultServerMux

## v0.2.1 (July 22, 2019)

BUG FIXES: 
- service: Force O gasPrice on readonly transactions (issue 5 of monetd)

## v0.2.0 (July 22, 2019)

SECURITY:

FEATURES:

IMPROVEMENTS:
- demo: Use evm-lite-lib package in demo scripts.
- state: Move genesis account creation from service to state. 
- state: PoA smart-contract bindings.

BUG FIXES:
- state: Initialize from empty state instead of latest trie root. This enables
         bootstrapping evm-lite/babble nodes from the babble DB only.

## V0.1.1 (January 28, 2019)

SECURITY:

FEATURES:

IMPROVEMENTS:
- deps: Use Geth v1.8.17
- consensus: Version 0.4.1 of Babble

BUG FIXES:
- deps: Update version of Geth. Version 1.7.0 had broken dependencies.

## V0.1.0 (October 14, 2018)

FEATURES:
- state: EVM, Trie, and Database.
- service: account management and HTTP API.
- consensus: simple consensus interface.
- consensus/solo: consensus implementation that simply relays transactions from
  service to state.
- consensus/babble: consensus implementation that uses an inmemory Babble node.
- consensus/raft: consensus implementation using hashicorp/raft
- engine: agent coordinating State, Service and Consensus.
- cmd: CLI
- deploy: scripts to create testnets locally or in AWS.


demo/src/poa.js
const util = require('util');
const fs = require('fs');
const JSONbig = require('json-bigint');
const argv = require('minimist')(process.argv.slice(2));
const evmlc = require('evm-lite-lib');
const solc = require('solc');
const prompt = require('prompt');

const FgRed = '\x1b[31m';
const FgGreen = '\x1b[32m';
const FgYellow = '\x1b[33m';
const FgBlue = '\x1b[34m';
const FgMagenta = '\x1b[35m';
const FgCyan = '\x1b[36m';
const FgWhite = '\x1b[37m';

const log = (color, text) => {
	console.log(color + text + '\x1b[0m');
};

const online = true;

const sleep = function(time) {
	return new Promise(resolve => setTimeout(resolve, time));
};

/**
 * Demo starts here.
 */

function Node(name, host, port) {
	this.name = name;
	this.api = new evmlc.EVMLC(host, port, {
		from: '',
		gas: 1000000,
		gasPrice: 0
	});
	this.account = {};
	this.ip = host;
}

const allAccounts = [];
const allNodes = [];

var crowdFunding = {};
var contractPath = '';
var nodeno = 0;


const init = async () => {
	console.group('Initialize Nodes: ');

	const ips = argv.ips 
		.replace(/\s/g, '')
		.split(',');
//		.sort();
	const port = argv.port;
	const keystore = new evmlc.Keystore(argv.keystore, 'keystore');
	const passwordPath = argv.pwd;
	nodeno = argv.nodeno;
	const password = readPasswordFile(passwordPath);

	contractPath = argv.contract;

	console.log('Sorted IPS: ', ips);
    console.log('Node No: ', nodeno);
	console.log('Keystore Path: ', keystore.path);
	console.log('Password File Path: ', passwordPath);
	console.log('Contract File Path: ', contractPath);

	for (i = 0; i < ips.length; i++) {
		node = new Node(util.format('node%s',nodeno ), ips[i], port);
		allNodes.push(node);
	}

	console.groupEnd();

	return {
		keystore,
		password
	};
};

const readPasswordFile = path => {
	return fs.readFileSync(path, { encoding: 'utf8' });
};

const decryptAccounts = async ({ keystore, password }) => {
	console.group('Decrypt Accounts');
	console.log('Password: ', password);

        const baseAccounts = await keystore.list(allNodes[0].api);

	for (const baseAccount of baseAccounts) {
		account = await keystore.decrypt(baseAccount.address, password);
		account.balance = baseAccount.balance;

		console.log('Decrypted: ', `${account.address}(${account.balance})`);
		allAccounts.push(account);
	}

	for (i = 0; i < allNodes.length; i++) {
		allNodes[i].api.defaultFrom = allAccounts[i].address;
		allNodes[i].account = allAccounts[i];
	}

	console.groupEnd();
};

const compiledSmartContract = async () => {
	const input = fs.readFileSync(contractPath, {
		encoding: 'utf8'
	});
	const output = solc.compile(input.toString(), 1);
	const bytecode = output.contracts[`:POA_Genesis`].bytecode;
	const abi = output.contracts[`:POA_Genesis`].interface;

	const contract = await allNodes[0].api.contracts.load(JSON.parse(abi), {
		data: bytecode
	});

	return contract;
};

class GenesisContract {
	constructor (account, ip) {
		this.account = account;
		this.contract = new evmlc.Contract({
            "gas":100000000, 
            "gasPrice": 0, 
            "from": this.account.address, 
            "address": "0xabbaabbaabbaabbaabbaabbaabbaabbaabbaabba", 
            "interface": [{"constant":true,"inputs":[{"name":"_publicKey","type":"bytes32"}],"name":"checkAuthorisedPublicKey","outputs":[{"name":"","type":"bool"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[{"name":"_address","type":"address"}],"name":"dev_isGenesisWhitelisted","outputs":[{"name":"","type":"bool"}],"payable":false,"stateMutability":"pure","type":"function"},{"constant":true,"inputs":[],"name":"dev_27","outputs":[{"name":"","type":"uint256"}],"payable":false,"stateMutability":"pure","type":"function"},{"constant":true,"inputs":[{"name":"_address","type":"address"}],"name":"checkAuthorised","outputs":[{"name":"","type":"bool"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[{"name":"","type":"address"}],"name":"whiteList","outputs":[{"name":"person","type":"address"},{"name":"flags","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[],"name":"dev_getSender","outputs":[{"name":"","type":"address"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[],"name":"dev_getWhitelistCount","outputs":[{"name":"","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":false,"inputs":[{"name":"_nomineeAddress","type":"address"},{"name":"_accepted","type":"bool"}],"name":"castNomineeVote","outputs":[{"name":"decided","type":"bool"},{"name":"voteresult","type":"bool"}],"payable":true,"stateMutability":"payable","type":"function"},{"constant":true,"inputs":[],"name":"dev_getGenesisWhitelist0","outputs":[{"name":"","type":"address"}],"payable":false,"stateMutability":"pure","type":"function"},{"constant":true,"inputs":[{"name":"_address","type":"address"}],"name":"dev_isWhitelisted","outputs":[{"name":"","type":"bool"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[{"name":"_address","type":"address"}],"name":"dev_getCurrentNomineeVotes","outputs":[{"name":"yes","type":"uint256"},{"name":"no","type":"uint256"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":false,"inputs":[{"name":"_nomineeAddress","type":"address"},{"name":"_moniker","type":"bytes32"}],"name":"submitNominee","outputs":[],"payable":true,"stateMutability":"payable","type":"function"},{"inputs":[{"name":"_moniker","type":"bytes32"}],"payable":false,"stateMutability":"nonpayable","type":"constructor"},{"anonymous":false,"inputs":[{"indexed":true,"name":"_nominee","type":"address"},{"indexed":false,"name":"_yesVotes","type":"uint256"},{"indexed":false,"name":"_noVotes","type":"uint256"},{"indexed":true,"name":"_accepted","type":"bool"}],"name":"NomineeDecision","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"name":"_nominee","type":"address"},{"indexed":true,"name":"_voter","type":"address"},{"indexed":false,"name":"_yesVotes","type":"uint256"},{"indexed":false,"name":"_noVotes","type":"uint256"},{"indexed":true,"name":"_accepted","type":"bool"}],"name":"NomineeVoteCast","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"name":"_nominee","type":"address"},{"indexed":true,"name":"_proposer","type":"address"}],"name":"NomineeProposed","type":"event"},{"anonymous":false,"inputs":[{"indexed":true,"name":"_address","type":"address"},{"indexed":true,"name":"_moniker","type":"bytes32"}],"name":"MonikerAnnounce","type":"event"}]
            },  ip , 8080 );
    }

    async genericTrans(functionname, value, params) {
        console.log(functionname, params);

        const transaction = await this.contract.methods[functionname](...params);
    
        transaction.value(value);

        console.log("From", transaction.tx.from) ;
        console.log("Data", transaction.tx.data) ;

        const response = await transaction.submit(this.account, { timeout: 3 });
  
        if (value > 0)
        {
           const receipt = await transaction.receipt;
  
           console.dir(receipt, { depth: 6, colors: true });
  
           const logs = this.contract.parseLogs(receipt.logs);
  
           for (const log of logs) {
               console.log(
                    log.event || 'No Event Name',
                    JSON.stringify(log.args, null, 2)
               );
           }
        }
        else
        {
           console.dir(response, { depth: 6, colors: true });
        }
        return transaction;
     }
}
	